### ISSUE TITLE
Make possible to run a linter without stdin or auto_append

### ISSUE THREAD

**Issue by kerma** [SublimeLinter/SublimeLinter#1520]
First, thank you for the great work. SublimeLinter is a very nice framework and pleasant to use, from the user and plugin developer perspective. 

I have a tiny issue with the API however. SublimeLinter provides a great framework to run some background jobs, however it's currently a little limited to using a linter either by sending code to it via stdin or running it against a (temporary) file. I have a third use case, where I'd like to run a linter on the working directory without specifying a filename or sending stdin to the command. For me to achieve what I want to do I need to override the `Linter.communicate` method. There is a side-effect there which decides based on the `code` variable whether to append file name or pass code via stdin. 

Passing an empty string via stdin may work, but it's not so nice and proper long term solution I guess. When overriding the communicate I'd still like to use `Linter._communicate` as it's doing what I need. But using an internal method is also not a good solution in long term.

So, I'm asking either:
a) Remove the side effect and make it possible for the caller (API user) to decide the file appending and stdin behaviour (with current defaults)
b) Make `Linter._communicate` to be public method (renaming the current `communicate` is probably not an option). In that case the plugin developer can use it safely. 

I could even do a PR for this if you agree to the proposal.

**Comment by kaste** [comment]
Hi @kerma!

You directly hit some long term issues.

Just to ensure, I understand you correctly. This boils down to just executing something like 

````
$ flake8 .
``` 

on the command line. The problem in running something like this is that it lints a whole folder or project, and usually reports errors from various locations (files). So to fully support this style of invocation we would need to allow that a single lint process can report problems for different files. We don't have this right now so you would have to filter the lint result to only report errors from one file. That's the 😢 thing.

Other than that you can probably achieve what you want here. Since you want to act on a folder (the working dir) you surely don't lint the actual buffers and thus can only lint when the file is saved and non dirty. Setting `tempfile_suffix = '-'` will do that. Now you just need to use a context variable in `cmd` to deactivate auto-appending the current filename.

TLDR;

```
cmd = 'flake8 ${folder}'
tempfile_suffix = '-'
```

(If you don't want the folder in the cmd, you could also use an undefined context variable like `${this_is_a_hack}`. IIRC we don't err in that case and just substitute with the empty space.)

**Comment by braver** [comment]
Just a quick reply, gotta run and will look at in more detail later.

SublimeLinter is designed to provide lint feedback while you're editing, on-the-fly, preferably for every change you make, and (therefore) on a file-by-file basis. We have some provisions for changing that behavior, but that's the core use case.
This rarely works for linters that want to work on directories. They usually need to grok more code and cannot operate with the performance required for on-the-fly linting. 

If you can run on-demand, you can look into build systems. You can run whatever you need to run for whatever context makes sense, and you will get feedback and error reports from that. 

Usually, this is where language services come in. I don't know what language you're working with, but LSP is usually much more powerful than SublimeLinter.

**Comment by kerma** [comment]
@kaste Yup, you understood correctly. `${this_is_a_hack}` is an interesting idea actually, it may work in my use case.

I don't necessarily see the problem (or 😢) in having to implement `split_match` myself to parse the results properly later (https://github.com/kerma/SublimeLinter-contrib-goflow/blob/master/linter.py#L41). 

@braver you are correct, I am trying to "abuse" SL to do what it wasn't meant to do. But only because I think it does some of the things better or in a way how I'd like the thing behave. Build systems and LSP have their place, but they require too much special configuration, they are missing some of the behavior or it just requires too much different plugins to achieve what I'm after. 

I do mostly Python and Go development, some JS/TS/React. And I've tried everything that's out there for those languages. And I still prefer to write my own "IDE" if you will. Python for example has a great set of plugins to automatically format code according to pep8, automatically fix imports etc. All of those are built by different people, following different conventions and using different ways how to ask for input and provide output. They work, but take a toll on your mental effort to constantly switch between (and like with everything, too much stuff will slow stuff down). 

With SL you've developed a decent set of defaults and behavior That Just Works™. And the fact that I could run `go fmt` on every save with 5 lines of linter definition was just too good to let go 😬 And adding `go test` with highlighting failed tests took almost no effort as well (https://github.com/kerma/SublimeLinter-contrib-goflow/blob/master/linter.py). 

I know I'm asking something that might not be the project goal, but at the same time I believe that a developer friendly, not so restrictive or less opinionated API might benefit the ecosystem in general. 

Thanks again for taking the time to respond.

**Comment by kaste** [comment]
'flexible' and 'loose coupled' are generally 👍  and as you can see we actually made progress here. E.g. you couldn't do this with SL3. I also used SublimeLinter for strange stuff bc it has a really good API to configure running arbitrary executables. And the backend is threaded.. If you look at the stand-alone formatters written for Sublime they often take hundreds of commits to get this right. 

To take this further, we just need dozens of written plugins that misuse SublimeLinter. You can't make API improvements out of thin air.

Please note that auto_append has already been deprecated. Use the hack I mentioned or use `finalize_cmd`

```
def finalize_cmd(self, cmd, ctx, **kw):
	return cmd
	# OR
	return super().finalize_cmd(cmd, ctx, auto_append=False)
```
